sms_logs/src/aws.ts

```diff
@@ -0,0 +1,40 @@
+import l from "aws-lambda";
+import { SecretsManager } from "@aws-sdk/client-secrets-manager";
+
+export function decodeEvent(
+  ev: l.CloudWatchLogsEvent
+): l.CloudWatchLogsDecodedData {
+  const data = ev.awslogs.data;
+  const log = Buffer.from(data, "base64").toString("utf8");
+  const logData: l.CloudWatchLogsDecodedData = JSON.parse(log);
+
+  return logData;
+}
+
+export async function createAlertsConfig() {
+  const client = new SecretsManager({ region: "us-east-1" });
+
+  const endpoints = await Promise.all(
+    ["alerts", "alerts-dev", "alerts-dev-green"].map(async (s) => {
+      const url = `https://${s}.mbtace.com/sms_failure`;
```

Comment:
The canonical domain for production is `alerts.mbta.com` — though this technically also works, I'm not sure we should rely on it.

Backing up a bit, shouldn't we configure the URL in the environment, and have (at least) separate "dev" and "prod" instances of this Lambda? Then we can test out changes in a deployed environment, before deploying them to production. (For this PR we don't have to think about the details of the environments; this script can just assume there will be an environment variable that contains the URL to report to.)

Reply:
hmmm... so the trick here is that I believe we can only have 2 "subscription filters" per log group. We're using up one of the slots for the existing Splunk forwarding, and the plan was for this lambda to use the other one (so we'd be temporarily maxing them out until this replaces the Splunk forwarding).

Given the above, I think we could go with an env var approach though it would get a bit messy. Any other thoughts on this?

Reply:
Then this limit seems like a blocker for being able to test new versions of the Lambda code in a production-like environment without actually deploying to production, which normally we place a pretty high value on. I will defer to @ianwestcott on this since I'm on the way out, but my initial reaction would be to look into alternate implementations that would allow us to maintain a dev/prod separation. (Ideally we'd have done so before writing all this code... and I'll definitely take the blame for that, since I approved of this approach in our earlier meeting without fully understanding the implications of the 2-filter limit.)

Our options are probably:

1. Proceed with the current Lambda approach, and just be really careful about testing and deploying changes in the future, since our only option is to go straight to production. In this case I'd recommend configuring the Lambda using an environment variable that holds a JSON-encoded object, which can then be as complex as needed to provide the multiple reporting URLs and corresponding secret names (the Lambda code would still have to use Secrets Manager to look up the values). Terraform has a `jsonencode` function which makes this very easy, and there is precedent with [Concentrate](https://github.com/mbta/devops/blob/master/terraform/prod/concentrate.tf).

2. Hook up the subscription filter to a Kinesis stream (which seems to be the only other option besides a Lambda). This allows using Firehose delivery streams to send the events directly to HTTP endpoints of our choosing, meaning no Lambda code would be required. There would be separate delivery streams per T-Alerts environment, but they could all pull from the same single Kinesis stream. Firehose does support configuring an arbitrary "access key" which is sent to the endpoint as an `X-Amz-Firehose-Access-Key` header, [as documented here](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html), so that would cover the authorization piece. There is some additional infra complexity due to the requirement to set up an S3 bucket/prefix where events go if HTTP delivery fails (you can't just drop them on the floor, even though we might be fine with that here).

Notably, Firehose also supports delivery to Splunk. However, the extra infra complexity and the fact that there's an account-wide soft limit of 50 Firehose delivery streams suggests that maybe we shouldn't aim to replace our existing Lambda-based Splunk forwarding with this approach.